clarify readme

Bruno-366 opened the pull request.

Bruno-366 committed:
add links to remark #startGroup/endGroup

Bruno-366 committed:
mention the other docker images

and clarify some of the language

Bruno-366 committed:
fix links to startGroup/endGroup headings

Bruno-366 committed:
add link to section about opam

erikmd commented on README.md:
```diff
@@ -24,8 +24,15 @@
+This is a GitHub action that uses (by default) 
+[coqorg/coq](https://hub.docker.com/r/coqorg/coq/) Docker images,
+which in turn is based on [coqorg/base](https://hub.docker.com/r/coqorg/coq/),
+a Docker image with a debian enviroment.
```
Typo:

```suggestion
a Docker image with a Debian environment.
```

erikmd commented on README.md:
```diff
@@ -24,8 +24,15 @@
+a Docker image with a debian enviroment.
+
+    docker-coq-action (GitHub action)
+    --> docker-coq (Dockerfile)
+    --> docker-base (Dockerfile still from coq-community)
+    --> debian docker
```
to be even more precise (if you also find it useful), I'd add both names of GiHub repo and Docker Hub repo:
```suggestion
    --> docker-coq (Dockerfile still in coq-community & Docker image in coqorg/coq from Docker Hub)
    --> docker-base (Dockerfile also in coq-community & Docker image in coqorg/base from Docker Hub)
    --> debian (Docker image from Docker Hub)
```

Bruno-366 replied:
yeah, I was thinking of doing that, but in a way that is still readable. The problem isn't so much that the information isn't there, rather that it becomes a lot of information at once for a newbie

erikmd replied:
OK @Bruno-366, indeed it's more readable with a table :)

erikmd commented on README.md:
```diff
@@ -70,7 +81,7 @@ steps:
 See also the [example repo](https://github.com/erikmd/docker-coq-github-action-demo).
 
-### Inputs
+### Inputs to the action.yaml file
```
Albeit this is correct, there is a possible confusion here between the `action.yaml` (which is part of the docker-coq-action repository and acts as an "internal metadata file for the GHA platform"), and the `workflow.yml` file (which is the only file intended to be written by users).

So maybe I'd suggest to write merely something like:

```suggestion
### Inputs to the docker-coq-action
```

Bruno-366 replied:
Hmm, I understand what you mean, but as there isn't any `workflow.yml` file in this repo, I suppose a user would base it on `action.yaml` given that slighty above (under usage) it says "see `action.yml`".

This is what I see on the Github action marketplace
```
- name: Docker-Coq Action
  # You may pin to the exact commit or the version.
  # uses: coq-community/docker-coq-action@5ede7d711a7eae792579832cca39133bbc9d2644
  uses: coq-community/docker-coq-action@v1.2.2
  with:
    # The path of the .opam file (or a directory), relative to the repo root.
    opam_file: # optional, default is .
    # "8.X", "latest", or "dev".
    coq_version: # optional, default is latest
    # minimal, 4.07-flambda, or 4.09-flambda
    ocaml_version: # optional, default is minimal
    # Customizable script run before "install", empty by default.
    before_install: # optional, default is startGroup "Print opam config"
  opam config list; opam repo list; opam list
endGroup

    # Customizable script to install the opam dependencies.
    install: # optional, default is startGroup "Install dependencies"
  opam pin add -n -y -k path $PACKAGE $WORKDIR
  opam update -y
  opam install -y -j 2 $PACKAGE --deps-only
endGroup

    # Customizable script run after "install".
    after_install: # optional, default is startGroup "List installed packages"
  opam list
endGroup

    # Customizable script run before the opam build, empty by default.
    before_script: # optional
    # Customizable script run to install the opam package(s).
    script: # optional, default is startGroup "Build"
  opam install -y -v -j 2 $PACKAGE
  opam list
endGroup

    # Customizable script run after the opam build, empty by default.
    after_script: # optional
    # Customizable script run to uninstall the opam package(s).
    uninstall: # optional, default is startGroup "Uninstallation test"
  opam remove $PACKAGE
endGroup

    # The main script run in the container; may be overridden.
    custom_script: # optional, default is {{before_install}}
{{install}}
{{after_install}}
{{before_script}}
{{script}}
{{after_script}}
{{uninstall}}

    # The name of the Docker image to pull; unset by default.
    custom_image: # optional
    # Space-separated list of env variables to export to the custom_script.
    export: # optional
```

erikmd replied:
what is the URL where you see this text as is?

because AFAIK, the standard GitHub marketplace URL is https://github.com/marketplace/actions/docker-coq-action and just renders the README.md in the first place. And the fact it mentions the `actions.yml` file was only a choice by mine but, admittedly is not very user-friendly (as I recall that `actions.yml` is a priori not required to be parsed by end-users, and definitely not intended to be edited by them): the README.md documentation should just be self-contained.

So **I keep my suggestion** of not writing `### Inputs to the action.yaml file` nor `### Inputs to the action.yml` but just `### Inputs to the docker-coq-action` or merely `### Inputs` if you prefer.

erikmd replied:
BTW, do you agree of adding somewhere the text I had suggested [in my main comment](https://github.com/coq-community/docker-coq-action/pull/56#pullrequestreview-652829775)? Namely:

*add some helpful text like this somewhere (e.g. just after `## Example`):*

> Using a [GitHub Action](https://docs.github.com/en/actions) in your GitHub repository amounts to committing a file `.github/workflows/your-workflow-name.yml`, e.g., `.github/workflows/build.yml`, containing (among others), a snippet such as:
> ```
> runs-on: ubuntu-latest  # container actions require GNU/Linux
> […]
>       ocaml_version: ${{ matrix.ocaml_version }}
> ```
> Each field can be customized, see [below](#inputs-to-the-docker-coq-action) for the documentation of those specific to the docker-coq-action, or the GitHub Actions official documentation for the [standard fields involved in workflows](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions).

erikmd replied:
> BTW, do you agree of adding somewhere the text I had suggested [in my main comment](https://github.com/coq-community/docker-coq-action/pull/56#pullrequestreview-652829775)? Namely:

Sorry @Bruno-366 I've just realized you already committed this :)

Bruno-366 replied:
If I go to the actions tab and click on any "set up this workflow" I get sent to a page where I am creating a new yml file under `.github/workflows/`, only difference is that to the right there is a tab where you can search actions from the market place. Type in docker-coq and it says "Copy and paste the following snippet into your .yml file." 

![Screenshot 2021-05-07 at 12 54 58](https://user-images.githubusercontent.com/81762173/117439800-969d0d80-af33-11eb-83af-28d1cfe56ec6.png)
![Screenshot 2021-05-07 at 12 55 07](https://user-images.githubusercontent.com/81762173/117439797-969d0d80-af33-11eb-81e6-765d4db7e64e.png)
![Screenshot 2021-05-07 at 12 55 18](https://user-images.githubusercontent.com/81762173/117439796-96047700-af33-11eb-840d-009b7fbbcd1a.png)
![Screenshot 2021-05-07 at 12 55 26](https://user-images.githubusercontent.com/81762173/117439791-94d34a00-af33-11eb-8b61-d363786a7a98.png)

Bruno-366 replied:
But given that the `yml` in the marketplace isn't the same as `action.yml`, I'll follow your suggestion.

erikmd reviewed (commented):
Hi @Bruno-366, thanks for opening this!

I know that your PR is still in Draft mode, but I already left a few comments :)

BTW regarding my review comment w.r.t. the section title with `action.yaml`, do you think you could also add a comment such as the following somewhere (e.g. just after `## Example`?) → feel free to rephrase anyway:

> Using a [GitHub Action](https://docs.github.com/en/actions) in your GitHub repository amounts to committing a file `.github/workflows/your-workflow-name.yml`, e.g., `.github/workflows/build.yml`, containing (among others), a snippet such as:
> ```
> runs-on: ubuntu-latest  # container actions require GNU/Linux
> […]
>       ocaml_version: ${{ matrix.ocaml_version }}
> ```
> Each field can be customized, see [below](#inputs-to-the-docker-coq-action) for the documentation of those specific to the docker-coq-action, or the GitHub Actions official documentation for the [standard fields involved in workflows](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions).

Bruno-366 committed:
debian -> Debian

Co-authored-by: Erik Martin-Dorel <erik@martin-dorel.org>

Bruno-366 committed:
create table for context with other dockerfiles

Bruno-366 committed:
make it easier to see which inputs are optional

move the "optional" to the headings so that when scrolling/browsing and getting a quick overview, one can see that all the inputs (so far) are optional. If there ever is a required input this would make finding it while scrolling easier too

Bruno-366 committed:
make it easier to skim through and find defaults

Question: what is the default to `export`?

Bruno-366 committed:
add link to timegroup.sh under heading: Remarks

Bruno-366 committed:
add bash syntax highlighting to the bash snippets

Bruno-366 commented:
two questions: what is default value for `export` since its optional, and what language is `custom_script` (the mustaches)? `custom_script` expands to a bash script, but its written in something else (given that mustaches aren't part of bash).

Bruno-366 committed:
add snippet from erikmd under example

Bruno-366 committed:
remove an unnecessary link

Bruno-366 committed:
add link to: standard fields involved in workflows

Bruno-366 committed:
add bash syntax highlighting to opam commands

Bruno-366 commented:
I think the remarks should __maybe__ be moved so that they are __before__ the inputs section, if reading them first makes understanding the inputs easier.

Bruno-366 commented:
Also the opam related information should have its own heading/section.
And maybe include a sentence like:
> `coq` is built on-top of `ocaml` and so `coq` projects use `ocaml`'s package manager (`opam`) to build themselves. This github action supports `opam` out of the box.

erikmd commented:
> I think the remarks should **maybe** be moved so that they are **before** the inputs section, if reading them first makes understanding the inputs easier.

Indeed, why not.

erikmd commented:
> Also the opam related information should have its own heading/section.

but there's already `#### OPAM` (?)

or maybe, you just mean move this section with a higher heading, such as `## OPAM` ?

> And maybe include a sentence like:
> 
> > `coq` is built on-top of `ocaml` and so `coq` projects use `ocaml`'s package manager (`opam`) to build themselves. This github action supports `opam` out of the box.

:+1: good idea; you can commit this as is :) (maybe just capitalizing `GitHub Action`)

erikmd commented on README.md:
```diff
@@ -68,19 +81,27 @@ steps:
-### Inputs
+### Inputs to the action.yaml file
+
+#### `opam_file` *Optional*
```
I agree with this change (but just to do the remark it will change the implied "permalinks"… 😅)

erikmd commented on README.md:
```diff
@@ -94,114 +115,122 @@ fi
+See [`custom_script`](#custom_script) and [startGroup/endGroup](#startGroupendGroup) for more details.
 
-#### `custom_script`
+#### `custom_script` *Optional* 
```
Same remarks as [above](https://github.com/coq-community/docker-coq-action/pull/56/files#r627811092) for other headings.

Would you agree to write this instead, for each heading you changed?

```suggestion
#### <a name="custom_script></a>`custom_script` *Optional*
```
The drawback is that each such heading will be accessible by two internal links, #custom_script-optional *and* #custom_script ; so I'm really not sure it's better…

Bruno-366 replied:
It's kind of ugly, another option is that the first line under each heading would have something like the following format:

```
[optional|required] default: 
```

erikmd replied:
Hi @Bruno-366, I agree that putting manually links `#### <a name="foo"></a>` everywhere is kind of ugly.
On the one hand, I liked the underlying idea of making the *Optional* keyword more visible.
On the other hand, *all* docker-coq-action fields are actually optional :wink:, and the current version of the README is broken in this PR (the internal links do not work anymore).
So I'd just suggest you rephrase all H4 headings by putting the ***Optional*** keyword like this:
```
#### `ocaml_version`

***Optional***

The version of OCaml.

[…]
```
Can you update the PR accordingly? and I'll merge it :relieved:

erikmd commented on README.md:
````diff
@@ -94,114 +115,122 @@ fi
 ```
 
-See [`custom_script`](#custom_script) for more details.
+See [`custom_script`](#custom_script) and [startGroup/endGroup](#startGroupendGroup) for more details.
````
But then, note that this link is now broken: it should be
```… [`custom_script`](#custom_script-optional) …```
and likewise for all other changed permalinks.

Otherwise, better solution: manually hardcode the links to preserve the permalinks. See [this comment](https://github.com/coq-community/docker-coq-action/pull/56/files#r627811880).

Bruno-366 committed:
change heading to just inputs

Bruno-366 committed:
mention default of export

Bruno-366 committed:
move opam section to top

probably makes it clearer that this github action main use case is probably building coq projects

erikmd commented on README.md:
```diff
@@ -24,16 +24,44 @@
+| x | docker-coq-action | GitHub action | <n/a>
+| ↳ | docker-coq        | Dockerfile    | coqorg/coq
+| ↳ | docker-base       | Dockerfile    | coqorg/base
+| ↳ | debian docker     | Docker image  | \_/debian
```
For uniformity, I'd remove the `docker` keyword here (albeit it isn't a GH repo, but the `Type` column precises it anyway):
```suggestion
| ↳ | Debian            | Docker image  | \_/debian
```
```

Bruno-366 replied:
Well the repo as far as i could tell is docker-library/official-images
https://github.com/docker-library/official-images/blob/master/library/debian

erikmd commented on README.md:
```diff
@@ -68,19 +96,27 @@ steps:
+or the GitHub Actions official documentation for the
+[standard fields involved in workflows](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions).
+
+See [action.yml](./action.yml)
```
Nitpick:
```suggestion
See [action.yml](./action.yml).
```

Bruno-366 committed:
add fullstop

Co-authored-by: Erik Martin-Dorel <erik@martin-dorel.org>

Bruno-366 committed:
Debian github

Co-authored-by: Erik Martin-Dorel <erik@martin-dorel.org>

Bruno-366 committed:
move "optional" out of headings

erikmd committed:
fix: copy-and-paste typo (s/coq/base/)

erikmd committed:
detail: trim spaces

erikmd committed:
detail: Add missing backquotes

erikmd committed:
Add remark and link about ($PACKAGE, $WORKDIR)

erikmd committed:
Make remarks a H2 heading

erikmd committed:
detail: URLs have changed for GitHub Actions doc

erikmd commented:
I took the opportunity to refine a few other details in the markdown documentation; will squash-merge now.
Thanks again, @Bruno-366 !

erikmd commented:
> just a tip: the table can be added to the other repos and the `x` on the first column moved to the relevant row.

FYI @Bruno-366, this is now done :)

* <https://github.com/coq-community/docker-base#readme>
* <https://github.com/coq-community/docker-coq#readme>
